### PR TITLE
Enable test_torch.py tests for BFloat16 on cuda

### PR DESCRIPTION
**Pull request by izdeby**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22605 Added .bfloat16()
* **#22428 Enable test_torch.py tests for BFloat16 on cuda**

Differential Revision: [D16083575](https://our.internmc.facebook.com/intern/diff/D16083575)

### CODE REVIEW

File: aten/src/ATen/preprocess_declarations.py [hunk 1]
```diff
@@ -76,9 +76,9 @@ def expand(types):
         if 'CPU' in backend_types:
             backend_types['CPU'].discard('BFloat16')
 
-    # TODO: remove this hack once support for a bfloat16 tensor for CUDA is enabled
-    if 'CUDA' in backend_types:
-        backend_types['CUDA'].discard('BFloat16')
+    if not option.get('cuda_bfloat16', False):
+        if 'CUDA' in backend_types:
+            backend_types['CUDA'].discard('BFloat16')
```
Comment (ezyang) [inline aten/src/ATen/preprocess_declarations.py]: Not really removed, eh? :)

**Comment by ezyang** [comment]
It's a pretty big patch but it all looks reasonable to me. Letting @gchanan take a look.

**Comment by izdeby** [comment]
> It's a pretty big patch but it all looks reasonable to me. Letting @gchanan take a look.

@ezyang, yeah i know its big. The problem is that there are a lot of inner dependencies from one thing to another. For example, i had to enable couple methods just to make test utils work with BFloat16 type as it is and without hacks.

File: aten/src/ATen/preprocess_declarations.py (hunk 1 above)
Comment (izdeby) [reply aten/src/ATen/preprocess_declarations.py]: its coming soon :)

File: aten/src/ATen/Declarations.cwrap [hunk 2]
```diff
@@ -6,6 +6,7 @@
   cpu_bool: True
   cuda_bool: True
   cpu_bfloat16: True
+  cuda_bfloat16: True
```
[the remark this answers is not in the document]
Comment (izdeby) [reply aten/src/ATen/Declarations.cwrap]: This PR is already big enough. More functionality will come in next PRs.

File: aten/src/ATen/native/cuda/BinaryOpsKernel.cu [hunk 3]
```diff
@@ -13,7 +13,7 @@
 namespace at { namespace native {
 
 void add_kernel_cuda(TensorIterator& iter, Scalar alpha_scalar) {
-  AT_DISPATCH_ALL_TYPES_AND(kHalf, iter.dtype(), "add_cuda", [&]() {
+  AT_DISPATCH_ALL_TYPES_AND2(kHalf, kBFloat16, iter.dtype(), "add_cuda", [&]() {
```
[the remark this answers is not in the document]
Comment (izdeby) [reply aten/src/ATen/native/cuda/BinaryOpsKernel.cu]: This is needed for the test/common_utils.py. The way we compare two tensors requires some math ops.

File: aten/src/THC/CMakeLists.txt [hunk 4]
```diff
@@ -26,6 +26,14 @@ foreach(THC_FILE TensorMathCompareT TensorMathCompare TensorMathReduce TensorMas
    LIST(APPEND extra_src "${CMAKE_CURRENT_SOURCE_DIR}/generated/THC${THC_FILE}Bool.cu")
 endforeach()
 
+foreach(THC_FILE TensorMathReduce TensorMathCompareT TensorMathCompare TensorMathPointwise)
```
[the remark this answers is not in the document]
Comment (izdeby) [reply aten/src/THC/CMakeLists.txt]: sorted

**Pull request closed without merging**